File: packages/cli/src/commands/collections.js
import { readFileSync, writeFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { parseArgs } from 'node:util';
import { loadAuth, serverFetch } from './auth.js';

export async function collectionsCommand(args, dataDir) {
  const subcommand = args[0];

  if (!subcommand || subcommand === '--help') {
    console.log(`
apiforge collections - Server-side collection management

USAGE:
  apiforge collections <subcommand> [options]

SUBCOMMANDS:
  list                  List saved collections on server
  show <id>             Show collection detail
  save                  Save a spec file as a collection
  delete <id>           Delete a collection
  pull <id>             Download collection to local file
  push <file>           Upload local spec file to server

OPTIONS:
  --name <name>         Collection name (for save/push)
  --file <path>         Spec file path (for save/push)
  --output <path>       Output file path (for pull)

EXAMPLES:
  apiforge collections list
  apiforge collections save --name "My API" --file ./swagger.json
  apiforge collections pull abc123 --output ./my-api.json
  apiforge collections push ./swagger.json --name "Studio API"
  apiforge collections delete abc123
`);
    return;
  }

  // All collection commands require auth
  const auth = loadAuth(dataDir);
  if (!auth?.token) {
    console.error('Error: Not logged in. Run `apiforge auth login` first.');
    process.exit(1);
  }

  switch (subcommand) {
    case 'list':
      await listCollections(dataDir);
      break;
    case 'show':
      await showCollection(args.slice(1), dataDir);
      break;
    case 'save':
    case 'push':
      await pushCollection(args.slice(1), dataDir);
      break;
    case 'pull':
      await pullCollection(args.slice(1), dataDir);
      break;
    case 'delete':
      await deleteCollection(args.slice(1), dataDir);
      break;
    default:
      console.error(`Unknown collections subcommand: ${subcommand}`);
      process.exit(1);
  }
}

async function listCollections(dataDir) {
  const { status, ok, data } = await serverFetch(dataDir, '/api/collections');

  if (!ok) {
    console.error(`Failed to list collections: ${data?.error || status}`);
    process.exit(1);
  }

  if (!data || data.length === 0) {
    console.log('No collections found.');
    return;
  }

  console.log(`Found ${data.length} collection(s):\n`);
  for (const col of data) {
    const updated = col.updated_at ? new Date(col.updated_at).toLocaleString() : '-';
    console.log(`  ${col.id}  ${col.name}  (updated: ${updated})`);
  }
}

async function showCollection(args, dataDir) {
  const id = args[0];
  if (!id) {
    console.error('Error: collection ID is required');
    console.error('Usage: apiforge collections show <id>');
    process.exit(1);
  }

  const { ok, data } = await serverFetch(dataDir, `/api/collections/${id}`);
  if (!ok) {
    console.error(`Failed to get collection: ${data?.error || 'Not found'}`);
    process.exit(1);
  }

  console.log(`Collection: ${data.name}`);
  console.log(`ID: ${data.id}`);
  console.log(`Created: ${data.created_at}`);
  console.log(`Updated: ${data.updated_at}`);

  if (data.spec) {
    try {
      const spec = JSON.parse(data.spec);
      const paths = Object.keys(spec.paths || {});
      console.log(`\nEndpoints: ${paths.length}`);
      for (const p of paths) {
        const methods = Object.keys(spec.paths[p]).filter(m => m !== 'parameters');
        for (const m of methods) {
          console.log(`  ${m.toUpperCase().padEnd(7)} ${p}`);
        }
      }
    } catch {
      console.log('\nSpec: (raw data available, use --output to save)');
    }
  }
}

async function pushCollection(args, dataDir) {
  const { values } = parseArgs({
    args,
    options: {
      name: { type: 'string' },
      file: { type: 'string' },
    },
  });

  const filePath = values.file;
  if (!filePath) {
    console.error('Error: --file <path> is required');
    console.error('Usage: apiforge collections push --file ./swagger.json --name "My API"');
    process.exit(1);
  }

  const absPath = resolve(filePath);
  let specContent;
  try {
    specContent = readFileSync(absPath, 'utf-8');
    JSON.parse(specContent); // validate JSON
  } catch (err) {
    console.error(`Error reading spec file: ${err.message}`);
    process.exit(1);
  }

  const name = values.name || guessName(specContent, absPath);

  console.log(`Uploading "${name}" to server...`);

  const { ok, data } = await serverFetch(dataDir, '/api/collections', {
    method: 'POST',
    body: JSON.stringify({ name, spec: specContent }),
  });

  if (!ok) {
    console.error(`Failed to save collection: ${data?.error || 'Unknown error'}`);
    process.exit(1);
  }

  console.log(`Collection saved: ${data.name} (ID: ${data.id})`);
}

async function pullCollection(args, dataDir) {
  const { values, positionals } = parseArgs({
    args,
    options: {
      output: { type: 'string', short: 'o' },
    },
    allowPositionals: true,
  });

  const id = positionals[0];
  if (!id) {
    console.error('Error: collection ID is required');
    console.error('Usage: apiforge collections pull <id> --output ./spec.json');
    process.exit(1);
  }

  const { ok, data } = await serverFetch(dataDir, `/api/collections/${id}`);
  if (!ok) {
    console.error(`Failed to get collection: ${data?.error || 'Not found'}`);
    process.exit(1);
  }

  const outputPath = values.output || `${data.name.replace(/[^a-zA-Z0-9-_]/g, '_')}.json`;
  const absOutput = resolve(outputPath);

  try {
    const spec = JSON.parse(data.spec);
    writeFileSync(absOutput, JSON.stringify(spec, null, 2));
  } catch {
    writeFileSync(absOutput, data.spec || '{}');
  }

  console.log(`Collection "${data.name}" saved to ${absOutput}`);
}

async function deleteCollection(args, dataDir) {
  const id = args[0];
  if (!id) {
    console.error('Error: collection ID is required');
    console.error('Usage: apiforge collections delete <id>');
    process.exit(1);
  }

  const { ok, data } = await serverFetch(dataDir, `/api/collections/${id}`, {
    method: 'DELETE',
  });

  if (!ok) {
    console.error(`Failed to delete: ${data?.error || 'Not found'}`);
    process.exit(1);
  }

  console.log(`Collection ${id} deleted.`);
}

function guessName(specContent, filePath) {
  try {
    const spec = JSON.parse(specContent);
    if (spec.info?.title) return spec.info.title;
  } catch { /* ignore */ }
  const base = filePath.split(/[/\\]/).pop();
  return base.replace(/\.json$/i, '');
}

File: packages/apiforge-rs/crates/apiforge-cli/src/commands/collections.rs
use anyhow::Result;
use apiforge_core::server;
use crossterm::style::Stylize;

use crate::CollectionsAction;

pub async fn execute(action: CollectionsAction) -> Result<()> {
    // All collection commands require auth
    let token = apiforge_core::auth::resolve_token();
    if token.is_none() {
        anyhow::bail!("Not logged in. Run `apiforge auth login` first.");
    }

    match action {
        CollectionsAction::List => {
            let resp = server::server_fetch("/api/collections", "GET", None).await?;
            if !resp.ok {
                let err = resp.body.get("error").and_then(|e| e.as_str()).unwrap_or("Unknown error");
                anyhow::bail!("Failed to list collections: {}", err);
            }

            let collections = resp.body.as_array();
            match collections {
                Some(arr) if !arr.is_empty() => {
                    println!("{}", format!("Found {} collection(s):", arr.len()).bold());
                    println!();
                    for col in arr {
                        let id = col.get("id").and_then(|i| i.as_str()).unwrap_or("-");
                        let name = col.get("name").and_then(|n| n.as_str()).unwrap_or("Untitled");
                        let updated = col.get("updated_at").and_then(|u| u.as_str()).unwrap_or("-");
                        println!("  {}  {}  (updated: {})", id.dark_grey(), name.bold(), updated);
                    }
                }
                _ => {
                    println!("{}", "No collections found.".dark_grey());
                }
            }
        }
        CollectionsAction::Show { id } => {
            let resp = server::server_fetch(&format!("/api/collections/{}", id), "GET", None).await?;
            if !resp.ok {
                let err = resp.body.get("error").and_then(|e| e.as_str()).unwrap_or("Not found");
                anyhow::bail!("Failed to get collection: {}", err);
            }

            let name = resp.body.get("name").and_then(|n| n.as_str()).unwrap_or("Untitled");
            let col_id = resp.body.get("id").and_then(|i| i.as_str()).unwrap_or("-");
            let created = resp.body.get("created_at").and_then(|c| c.as_str()).unwrap_or("-");
            let updated = resp.body.get("updated_at").and_then(|u| u.as_str()).unwrap_or("-");

            println!("{}", format!("Collection: {}", name).bold());
            println!("ID: {}", col_id);
            println!("Created: {}", created);
            println!("Updated: {}", updated);

            // Try to parse spec and show endpoints
            if let Some(spec_str) = resp.body.get("spec").and_then(|s| s.as_str()) {
                if let Ok(spec) = serde_json::from_str::<serde_json::Value>(spec_str) {
                    if let Some(paths) = spec.get("paths").and_then(|p| p.as_object()) {
                        println!();
                        println!("Endpoints: {}", paths.len().to_string().cyan());
                        for (path, methods) in paths {
                            if let Some(methods) = methods.as_object() {
                                for (method, _) in methods {
                                    if method == "parameters" {
                                        continue;
                                    }
                                    println!("  {} {}", method.to_uppercase().bold(), path);
                                }
                            }
                        }
                    }
                }
            }
        }
        CollectionsAction::Push { file, name } => {
            let abs_path = std::fs::canonicalize(&file)
                .map_err(|e| anyhow::anyhow!("Failed to resolve path {}: {}", file, e))?;
            let spec_content = std::fs::read_to_string(&abs_path)
                .map_err(|e| anyhow::anyhow!("Failed to read {}: {}", file, e))?;

            // Validate JSON
            let _: serde_json::Value = serde_json::from_str(&spec_content)
                .map_err(|e| anyhow::anyhow!("Invalid JSON: {}", e))?;

            let collection_name = name.unwrap_or_else(|| {
                // Try to extract title from spec
                if let Ok(spec) = serde_json::from_str::<serde_json::Value>(&spec_content) {
                    if let Some(title) = spec.get("info").and_then(|i| i.get("title")).and_then(|t| t.as_str()) {
                        return title.to_string();
                    }
                }
                abs_path.file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("Untitled")
                    .to_string()
            });

            println!("Uploading \"{}\" to server...", collection_name);

            let resp = server::server_fetch(
                "/api/collections",
                "POST",
                Some(serde_json::json!({
                    "name": collection_name,
                    "spec": spec_content,
                })),
            ).await?;

            if !resp.ok {
                let err = resp.body.get("error").and_then(|e| e.as_str()).unwrap_or("Unknown error");
                anyhow::bail!("Failed to save collection: {}", err);
            }

            let id = resp.body.get("id").and_then(|i| i.as_str()).unwrap_or("-");
            let saved_name = resp.body.get("name").and_then(|n| n.as_str()).unwrap_or(&collection_name);
            println!("{}", format!("Collection saved: {} (ID: {})", saved_name, id).green());
        }
        CollectionsAction::Pull { id, output } => {
            let resp = server::server_fetch(&format!("/api/collections/{}", id), "GET", None).await?;
            if !resp.ok {
                let err = resp.body.get("error").and_then(|e| e.as_str()).unwrap_or("Not found");
                anyhow::bail!("Failed to get collection: {}", err);
            }

            let name = resp.body.get("name").and_then(|n| n.as_str()).unwrap_or("collection");
            let output_path = output.unwrap_or_else(|| {
                let safe_name: String = name.chars()
                    .map(|c| if c.is_alphanumeric() || c == '-' || c == '_' { c } else { '_' })
                    .collect();
                format!("{}.json", safe_name)
            });

            let spec_str = resp.body.get("spec").and_then(|s| s.as_str()).unwrap_or("{}");
            let content = match serde_json::from_str::<serde_json::Value>(spec_str) {
                Ok(spec) => serde_json::to_string_pretty(&spec)?,
                Err(_) => spec_str.to_string(),
            };

            std::fs::write(&output_path, &content)?;
            println!("{}", format!("Collection \"{}\" saved to {}", name, output_path).green());
        }
        CollectionsAction::Delete { id } => {
            let resp = server::server_fetch(
                &format!("/api/collections/{}", id),
                "DELETE",
                None,
            ).await?;

            if !resp.ok {
                let err = resp.body.get("error").and_then(|e| e.as_str()).unwrap_or("Not found");
                anyhow::bail!("Failed to delete: {}", err);
            }

            println!("{}", format!("Collection {} deleted.", id).green());
        }
    }
    Ok(())
}
